### PR TITLE
SDNA Active record implementation

### PR DESCRIPTION
**Pull request by fayeed**

### CODE REVIEW

File: core/src/subject/SubjectEntity.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+export class SubjectEntity {
+  #baseExpression: string;
+  #subjectClass: string;
+  #perspective: PerspectiveProxy
+  author: string;
+  timestamp: string;
+
+  constructor(perspective: PerspectiveProxy, baseExpression?: string) {
+    this.#baseExpression = baseExpression ? baseExpression : `entry://${makeRandomPrologAtom(24)}`;
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]:
```suggestion
    this.#baseExpression = baseExpression ? baseExpression : Literal.from(makeRandomPrologAtom(24)).toUrl();
```

File: core/src/subject/SubjectEntity.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  private async getData(id?: string) {
+    const tempId = id ?? this.#baseExpression;
+    let isInstance = await this.#perspective.isSubjectInstance(tempId, this.#subjectClass)
+    if (!isInstance) {
+      throw `Not a valid subject instance of ${this.#subjectClass} for ${tempId}`
+    }
+
+    const subject = await this.#perspective.getSubjectProxy(tempId, this.#subjectClass);
+
+    if (!subject) {
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: Don't really understand what this is for.. ?
`subject` isn't used below. if the base is not an instance of the class (which createSubject() would fix) we already jumped out of this function with the throw above...

File: core/src/subject/SubjectEntity.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+      await this.#perspective.createSubject(this.#subjectClass, tempId);
+    }
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: I think these lines can go
```suggestion
```

File: core/src/subject/SubjectEntity.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+      const resolveExpressionURI = await this.#perspective.infer(`subject_class("${this.#subjectClass}", C), property_resolve(C, "${p}")`)
+      const getProperty = async () => {
+        let results = await this.#perspective.infer(`subject_class("${this.#subjectClass}", C), property_getter(C, "${tempId}", "${p}", Value)`)
+        if (results && results.length > 0) {
+          let expressionURI = results[0].Value
+
+          if (resolveExpressionURI) {
+            const expression = await this.#perspective.getExpression(expressionURI)
+            this['author'] = expression.author;
+            this['timestamp'] = expression.timestamp;
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: That doesn't seem right. The expression we're setting in a property could be created by somebody else an definitely at a different timestamp than the base..

File: core/src/subject/SubjectEntity.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+      this[c] = await getProperty()
+    }
+
+    this.#baseExpression = tempId;
+
+    return this
+  }
+
+  private async setProperties(key: string, value: any) {
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: This only sets one property, right?
```suggestion
  private async setProperty(key: string, value: any) {
```

File: core/src/subject/SubjectEntity.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    const setters = await this.#perspective.infer(`subject_class("${this.#subjectClass}", C), property_setter(C, Property, Setter)`)
+    if (setters.length > 0) {
+      const setter = setters.find(e => e.Property === key)
+
+      if (setter) {
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: Why not let Prolog do the matching and only look at the given property?
```js
await this.#perspective.infer(`subject_class("${this.#subjectClass}", C), property_setter(C, ${key}, Setter)`)
```

File: core/src/subject/SubjectEntity.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+          await this.#perspective.executeAction(actions, this.#baseExpression, [{ name: "value", value }])
+        }
+      }
+    }
+  }
+
+  async save() {
+    this.#subjectClass = await this.#perspective.stringOrTemplateObjectToSubjectClass(this)
+    
+    const subject = await this.#perspective.createSubject(this, this.#baseExpression);
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]:
```suggestion
```

File: core/src/subject/SubjectEntity.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            break;
+          default:
+            await this.setCollectionSetter(key, value.value)
+            break;
+        }
+      } else {
+        await this.setCollectionSetter(key, value)
+      }
+    }
+  }
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: update() and save() are the same, right?

File: tests/js/integration.test.ts [hunk 9]
```diff
@@ -584,6 +586,73 @@ describe("Integration", () => {
...
+                    let root = Literal.from("Active record implementation test").toUrl()
+                    const recipe = new Recipe(perspective!, root)
+
+                    recipe.name = "recipe://test1";
+
+                    const recipes = await recipe.find();
+
+                    expect(recipes.length).to.equal(1)
+                })
+            })
```
Comment (lucksus) [inline tests/js/integration.test.ts:708]: Needs tests for collections. Would suggest to clone the collection tests from the subject proxy above.

**Review by lucksus — CHANGES_REQUESTED** [review]

File: core/src/subject/SubjectEntity.ts (hunk 7 above)
Comment (fayeed) [reply core/src/subject/SubjectEntity.ts]: The find all doesn't work if I don't create a subject instance for it.

File: core/src/subject/SubjectEntity.ts [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+  async get() {
+    this.#subjectClass = await this.#perspective.stringOrTemplateObjectToSubjectClass(this)
+    
+    return await this.getData()
+  }
+
+  // TODO: implement simple quering like limit, skip etc.
+  async find() {
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]: Find it weird that this is an instance method, and not a static class method. I understand that we need a PerspectiveProxy to do anything, but we can provide that as argument. Clearer interface than having to create an instance which is not used as an instance of the Subject Class..
```suggestion
  static async all(perspective: PerspectiveProxy) {
```

File: core/src/subject/SubjectEntity.ts [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+    this.#subjectClass = await this.#perspective.stringOrTemplateObjectToSubjectClass(this)
+    const proxies = await this.#perspective.getAllSubjectProxies(this.#subjectClass)
```
Comment (lucksus) [inline core/src/subject/SubjectEntity.ts]:
```suggestion
    let subjectClass = await perspective.stringOrTemplateObjectToSubjectClass(this)
    const proxies = await perspective.getAllSubjectProxies(subjectClass)
```

File: tests/js/integration.test.ts [hunk 12]
```diff
@@ -584,6 +586,129 @@ describe("Integration", () => {
...
+
+                    expect(recipe2.name).to.equal("recipe://test1")
+                })
+
+                it("find()", async () => {
+                    let root = Literal.from("Active record implementation test").toUrl()
+                    const recipe = new Recipe(perspective!, root)
+
+                    const recipes = await recipe.find();
+
```
Comment (lucksus) [inline tests/js/integration.test.ts]: With above changes, this becomes much clearer...
```suggestion
                    const recipes = new Recipe.all(perspective!)

```

**Review by lucksus — CHANGES_REQUESTED** [review]
Would like `find()` to be changed to `static all()` as described below... Rest is great!

**Review by lucksus — APPROVED** [review]
👍

**Pull request merged**